receiver/kafkametricsreceiver/consumer_scraper.go

```diff
@@ -0,0 +1,179 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkametricsreceiver
+
+import (
+	"context"
+	"regexp"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/simple"
+	"go.opentelemetry.io/collector/receiver/scrapererror"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkametricsreceiver/internal/metadata"
+)
+
+type consumerScraper struct {
+	client       sarama.Client
+	logger       *zap.Logger
+	groupFilter  *regexp.Regexp
+	topicFilter  *regexp.Regexp
+	clusterAdmin sarama.ClusterAdmin
+}
+
+func (s *consumerScraper) Name() string {
+	return "consumers"
+}
+
+func (s *consumerScraper) shutdown(_ context.Context) error {
+	if !s.client.Closed() {
+		return s.client.Close()
+	}
+	return nil
+}
+
+func (s *consumerScraper) scrape(context.Context) (pdata.ResourceMetricsSlice, error) {
+	metrics := simple.Metrics{
+		Metrics:                    pdata.NewMetrics(),
+		Timestamp:                  time.Now(),
+		MetricFactoriesByName:      metadata.M.FactoriesByName(),
+		InstrumentationLibraryName: InstrumentationLibName,
+	}
+
+	cgs, listErr := s.clusterAdmin.ListConsumerGroups()
+	if listErr != nil {
+		return metrics.ResourceMetrics(), listErr
+	}
+
+	var matchedGrpIds []string
+	for grpID := range cgs {
+		if s.groupFilter.MatchString(grpID) {
+			matchedGrpIds = append(matchedGrpIds, grpID)
+		}
+	}
+
+	allTopics, listErr := s.clusterAdmin.ListTopics()
+	if listErr != nil {
+		return metrics.ResourceMetrics(), listErr
+	}
+
+	matchedTopics := make(map[string]sarama.TopicDetail)
+	for t, d := range allTopics {
+		if s.topicFilter.MatchString(t) {
+			matchedTopics[t] = d
+		}
+	}
+	var scrapeErrors = scrapererror.ScrapeErrors{}
+	// partitionIds in matchedTopics
+	topicPartitions := make(map[string][]int32)
+	// currentOffset for each partition in matchedTopics
+	topicPartitionOffset := make(map[string]map[int32]int64)
+	for topic := range matchedTopics {
+		topicPartitions[topic] = make([]int32, 0)
+		topicPartitionOffset[topic] = make(map[int32]int64)
```

Comment:
Is it necessary since if you don't set it `topicPartitions[topic]` will return nil which is ok if you're doing `append(nil, p)` later.

Reply:
true, L#88 is not necessary - I forgot `append` works on nil slices.